src/DataProtection/DataProtection/src/KeyManagement/KeyRingBasedDataProtector.cs

```diff
@@ -14,8 +14,233 @@
 
 namespace Microsoft.AspNetCore.DataProtection.KeyManagement
 {
+#if NETCOREAPP
+    internal unsafe sealed class KeyRingBasedDataProtector : ISpanDataProtector, IPersistedDataProtector
+    {
+        public Span<byte> Protect(ReadOnlySpan<byte> plaintext)
+        {
+            if (plaintext == null)
+            {
+                throw new ArgumentNullException(nameof(plaintext));
+            }
+
+            try
+            {
+                // Perform the encryption operation using the current default encryptor.
+                var currentKeyRing = _keyRingProvider.GetCurrentKeyRing();
+                var defaultKeyId = currentKeyRing.DefaultKeyId;
+                var defaultEncryptorInstance = currentKeyRing.DefaultAuthenticatedEncryptor;
+                CryptoUtil.Assert(defaultEncryptorInstance != null, "defaultEncryptorInstance != null");
+
+                if (_logger.IsDebugLevelEnabled())
+                {
+                    _logger.PerformingProtectOperationToKeyWithPurposes(defaultKeyId, JoinPurposesForLog(Purposes));
+                }
+
+                // We'll need to apply the default key id to the template if it hasn't already been applied.
+                // If the default key id has been updated since the last call to Protect, also write back the updated template.
+                var aad = _aadTemplate.GetAadForKey(defaultKeyId, isProtecting: true);
+
+                Span<byte> retVal;
+                var optimizedEncryptor = defaultEncryptorInstance as ISpanAuthenticatedEncryptor;
+                var magicHeaderAndKeyOffset = sizeof(uint) + sizeof(Guid);
+                if (optimizedEncryptor != null)
+                {
+                    retVal = new byte[magicHeaderAndKeyOffset + plaintext.Length];
+                    optimizedEncryptor.Encrypt(retVal.Slice(magicHeaderAndKeyOffset), plaintext, aad);
+                }
+                else
+                {
+                    // We allocate a 20-byte pre-buffer so that we can inject the magic header and key id into the return value.
+                    retVal = defaultEncryptorInstance.Encrypt(
+                        plaintext: new ArraySegment<byte>(plaintext.ToArray()),
+                        additionalAuthenticatedData: new ArraySegment<byte>(aad),
+                        preBufferSize: (uint)(sizeof(uint) + sizeof(Guid)),
+                        postBufferSize: 0).AsSpan();
+                }
+
+                CryptoUtil.Assert(retVal != null && retVal.Length >= sizeof(uint) + sizeof(Guid), "retVal != null && retVal.Length >= sizeof(uint) + sizeof(Guid)");
+
+                // At this point: retVal := { 000..000 || encryptorSpecificProtectedPayload },
+                // where 000..000 is a placeholder for our magic header and key id.
+
+                // Write out the magic header and key id
+                fixed (byte* pbRetVal = retVal)
+                {
+                    WriteBigEndianInteger(pbRetVal, MAGIC_HEADER_V0);
```

Comment:
Theres' span APIs for this: https://docs.microsoft.com/en-us/dotnet/api/system.buffers.binary.binaryprimitives.writeint32bigendian?view=net-5.0